frontends/parsers/parserDriver.cpp

```diff
@@ -0,0 +1,238 @@
+    Util::InputSources::instance->appendText(text);
+    auto posAfterToken = Util::InputSources::instance->getCurrentPosition();
+    yylloc = Util::SourceInfo(posBeforeToken, posAfterToken);
+}
+
+void AbstractParserDriver::onReadLineNumber(const char* text) {
+    char* last;
+    errno = 0;
+    lineDirectiveLine = strtol(text, &last, 10);
+    if (errno != 0 ||
+        // we have not parsed the complete string
+        strlen(last) != 0)
+        ErrorReporter::instance.parser_error("Error parsing line number %s", text);
+}
+
+void AbstractParserDriver::onReadFileName(const char* text) {
+    lineDirectiveFile = cstring(text);
+    Util::InputSources::instance->mapLine(lineDirectiveFile, lineDirectiveLine);
+}
+
+void AbstractParserDriver::onReadIdentifier(cstring id) {
+    lastIdentifier = id;
+}
+
+void AbstractParserDriver::onParseError(const Util::SourceInfo& location,
+                                        const std::string& message) {
+    static const std::string unexpectedIdentifierError =
+        "syntax error, unexpected IDENTIFIER";
+    if (boost::equal(message, unexpectedIdentifierError)) {
+        ErrorReporter::instance.parser_error(location, boost::format("%s \"%s\"") %
+                                                       unexpectedIdentifierError %
+                                                       lastIdentifier);
+    } else {
+        ErrorReporter::instance.parser_error(location, message);
+    }
+}
+
+P4ParserDriver::P4ParserDriver()
+  : structure(new Util::ProgramStructure)
+  , declarations(new IR::IndexedVector<IR::Node>())
+{ }
+
+/* static */ const IR::P4Program*
+P4ParserDriver::parse(const char* name, std::istream& in) {
+    if (Log::verbose())
+        std::cout << "Parsing P4-16 program " << name << std::endl;
+
+    // Create and configure the parser and lexer.
+    P4ParserDriver driver;
+    P4Lexer lexer(in);
+    P4Parser parser(driver, lexer);
+
+#ifdef YYDEBUG
+    if (const char *p = getenv("YYDEBUG"))
+        parser.set_debug_level(atoi(p));
+    driver.structure->setDebug(parser.debug_level() != 0);
+#endif
+
+    // Parse.
+    if (parser.parse() != 0) return nullptr;
+    driver.structure->endParse();
+    return new IR::P4Program(driver.declarations->srcInfo, *driver.declarations);
+}
+
+/* static */ const IR::P4Program*
+P4ParserDriver::parse(const char* name, FILE* in) {
+    AutoStdioInputStream inputStream(in);
+    return parse(name, inputStream.get());
+}
+
+/* static */ void P4ParserDriver::checkShift(const Util::SourceInfo& l,
+                                                 const Util::SourceInfo& r) {
+    if (!l.isValid() || !r.isValid())
+        BUG("Source position not available!");
+    const Util::SourcePosition& f = l.getStart();
+    const Util::SourcePosition& s = r.getStart();
+    if (f.getLineNumber() != s.getLineNumber() ||
+        f.getColumnNumber() != s.getColumnNumber() - 1)
+        ::error("Syntax error at shift operator: %1%", l);
```

Comment:
I wonder whether this is a good error; it's not really a shift since there are spaces in the middle.
Anyway, I suspect this message is inherited from the old code.